// @storybook/cli/dist/cjs/generators/configure.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.configure = configure;

var _fsExtra = _interopRequireDefault(require("fs-extra"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function configureMain(_ref) {
  let {
    addons,
    extensions = ['js', 'jsx', 'ts', 'tsx'],
    commonJs = false
  } = _ref,
      custom = _objectWithoutPropertiesLoose(_ref, ["addons", "extensions", "commonJs"]);

  const prefix = _fsExtra.default.existsSync('./src') ? '../src' : '../stories';
  const config = Object.assign({
    stories: [`${prefix}/**/*.stories.mdx`, `${prefix}/**/*.stories.@(${extensions.join('|')})`],
    addons
  }, custom); // replace escaped values and delimiters

  const stringified = `module.exports = ${JSON.stringify(config, null, 2).replace(/\\"/g, '"').replace(/['"]%%/g, '').replace(/%%['"]/, '')}`;

  _fsExtra.default.ensureDirSync('./.storybook');

  _fsExtra.default.writeFileSync(`./.storybook/main.${commonJs ? 'cjs' : 'js'}`, stringified, {
    encoding: 'utf8'
  });
}

function configurePreview(framework, commonJs) {
  const parameters = `
export const parameters = {
  actions: { argTypesRegex: "^on[A-Z].*" },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
}`;
  const preview = framework === 'angular' ? `
import { setCompodocJson } from "@storybook/addon-docs/angular";
import docJson from "../documentation.json";
setCompodocJson(docJson);

${parameters}` : parameters;

  _fsExtra.default.writeFileSync(`./.storybook/preview.${commonJs ? 'cjs' : 'js'}`, preview, {
    encoding: 'utf8'
  });
}

function configure(framework, mainOptions) {
  _fsExtra.default.ensureDirSync('./.storybook');

  configureMain(mainOptions);
  configurePreview(framework, mainOptions.commonJs);
}
